crates/turborepo-api-client/src/lib.rs

```diff
@@ -240,7 +255,162 @@ impl APIClient {
+            request_builder = request_builder.header("x-artifact-tag", tag);
+        }
+
+        retry::make_retryable_request(request_builder)
+            .await?
+            .error_for_status()?;
+
+        Ok(())
+    }
+
+    pub async fn fetch_artifact(
+        &self,
+        hash: &str,
+        token: &str,
+        team_id: &str,
+        team_slug: Option<&str>,
+        use_preflight: bool,
+    ) -> Result<Response> {
+        self.get_artifact(hash, token, team_id, team_slug, use_preflight, Method::GET)
+            .await
+    }
+
+    pub async fn artifact_exists(
+        &self,
+        hash: &str,
+        token: &str,
+        team_id: &str,
+        team_slug: Option<&str>,
+        use_preflight: bool,
+    ) -> Result<Response> {
+        self.get_artifact(hash, token, team_id, team_slug, use_preflight, Method::HEAD)
+            .await
+    }
+
+    async fn get_artifact(
+        &self,
+        hash: &str,
+        token: &str,
+        team_id: &str,
+        team_slug: Option<&str>,
+        use_preflight: bool,
+        method: Method,
+    ) -> Result<Response> {
+        let mut request_url = self.make_url(&format!("/v8/artifacts/{}", hash));
+        let mut allow_auth = true;
+
+        if use_preflight {
+            let preflight_response = self
+                .do_preflight(token, &request_url, "GET", "Authorization, User-Agent")
+                .await?;
+
+            allow_auth = preflight_response.allow_auth;
+            request_url = preflight_response.location.to_string();
+        };
+
+        let mut request_builder = self
+            .client
+            .request(method, request_url)
+            .header("User-Agent", self.user_agent.clone());
+
+        if allow_auth {
+            request_builder = request_builder.header("Authorization", format!("Bearer {}", token));
+        }
+
+        request_builder = Self::add_team_params(request_builder, team_id, team_slug);
+
+        let response = retry::make_retryable_request(request_builder)
+            .await?
+            .error_for_status()?;
+
+        Ok(response)
+    }
+
+    pub async fn do_preflight(
+        &self,
+        token: &str,
+        request_url: &str,
+        request_method: &str,
+        request_headers: &str,
+    ) -> Result<PreflightResponse> {
+        let request_builder = self
+            .client
+            .request(Method::OPTIONS, request_url)
+            .header("User-Agent", self.user_agent.clone())
+            .header("Access-Control-Request-Method", request_method)
+            .header("Access-Control-Request-Headers", request_headers)
+            .header("Authorization", format!("Bearer {}", token));
+
+        let response = retry::make_retryable_request(request_builder).await?;
+
+        let headers = response.headers();
+        let location = if let Some(location) = headers.get("Location") {
+            let location = location.to_str()?;
+            Url::parse(location)?
```

Comment:
This parses the location as a fully-qualified domain. I suspect we should be using `join` here.

One test, please!

Reply:
This is also the same behavior as the Go code, should we change or keep the same?

Reply:
That's a bug, we should fix it. Since I think it is just `Url::join` vs. `Url::parse` let's bundle it.